File: 0001-0500/0092_reverseBetween.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        position = 1
        front = None
        curr = head
        while position < left:
            front = curr
            curr = curr.next
            position += 1
        back = curr
        if not curr or not curr.next: return head
        nxt = curr.next
        while position < right:
            if nxt.next:
                temp = nxt.next
            else:
                temp = None
            nxt.next = curr
            curr = nxt
            nxt = temp
            position += 1
        if front:
            front.next = curr
        else:
            head = curr
        back.next = nxt
        return head

# Alternative solution

class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        currPosition = 1
        tailFromLeft = None
        reverseHead = None
        temp = head
        while currPosition < left:
            tailFromLeft = temp
            currPosition += 1
            temp = temp.next
        reverseHead = temp
        prev, curr = None, reverseHead
        while currPosition <= right:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
            currPosition += 1
        if tailFromLeft:
            tailFromLeft.next = prev
        else:
            head = prev
        reverseHead.next = curr
        return head

File: 0001-0500/0092_reverseBetween.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode firstTail = null;
        ListNode firstHead = null;
        ListNode secondTail = null;
        ListNode secondHead = null;
        ListNode prev = null;
        ListNode curr = head;
        int i = 1;

        while (curr != null) {
            if (i == left) {
                firstTail = prev;
                firstHead = curr;
            } 
            if (i == right) {
                secondTail = curr;
                secondHead = curr.next;
            }
            i++;
            prev = curr;
            curr = curr.next;
        }

        secondTail.next = null;
        ListNode reversed = reverseList(firstHead);

        if (firstTail != null) firstTail.next = reversed;
        else head = reversed;
        if (firstHead != secondHead) firstHead.next = secondHead;

        return head;
    }

    public ListNode reverseList(ListNode node) {
        ListNode prev = null;
        ListNode curr = node;
        ListNode next;

        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
}

// Alternative solution

class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        ListNode firstTail = null;
        ListNode firstHead = head;

        int current = 1;

        while (firstHead != null) {
            if (current < left ) {
                firstTail = firstHead;
                firstHead = firstHead.next;
            } else {
                break;
            }
            current++;
        }

        ListNode secondTail = firstTail;
        ListNode secondHead = firstHead;

        while (secondHead != null) {
            if (current <= right) {
                secondTail = secondHead;
                secondHead = secondHead.next;
            } else {
                break;
            }
            current++;
        }



        secondTail.next = null;
        ListNode reversed = reverseList(firstHead);
        if (firstTail != null) {
            firstTail.next = reversed;
        } else {
            head = reversed;
        }
        if (firstHead != secondHead) {
            firstHead.next = secondHead;
        }
        return head;
    }

    public ListNode reverseList(ListNode node) {
        ListNode prev = null;
        ListNode curr = node;
        ListNode next;

        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        return prev;
    }
}
